imod/mf6/package.py

```diff
@@ -526,39 +526,56 @@ def mask(self, domain: GridDataArray) -> Any:
 
         Parameters
         ----------
-        domain: xr.DataArray of integers. Preservers values where domain is larger than 0.
+        mask: xr.DataArray, xu.UgridDataArray of ints
+            idomain-like integer array. 1 sets cells to active, 0 sets cells to inactive, 
+            -1 sets cells to vertical passthrough
 
         Returns
         -------
         masked: Package
             The package with part masked.
         """
+
         masked = {}
         for var in self.dataset.data_vars.keys():
-            da = self.dataset[var]
-            if self.skip_masking_dataarray(var):
-                masked[var] = da
-                continue
-            if set(domain.dims).issubset(da.dims):
-                if issubclass(da.dtype.type, numbers.Integral):
-                    masked[var] = da.where(domain > 0, other=0)
-                elif issubclass(da.dtype.type, numbers.Real):
-                    masked[var] = da.where(domain > 0)
-                else:
-                    raise TypeError(
-                        f"Expected dtype float or integer. Received instead: {da.dtype}"
-                    )
+            if self._is_copy_variable_on_masking(var, self.dataset[var]):
+                masked[var] = self.dataset[var]
             else:
-                if da.values[()] is not None:
-                    if is_scalar(da.values[()]):
-                        masked[var] = da.values[()]  # For scalars, such as options
-                    else:  # For example for arrays with only a layer dimension
-                        masked[var] = da
-                else:
-                    masked[var] = None
+                masked[var] = self._mask_spatial_var(var, mask)
 
         return type(self)(**masked)
 
+    def _is_copy_variable_on_masking(self, var: str, da: GridDataArray)->bool: 
+        if self.skip_masking_dataarray(var) or len(da.dims) == 0 or set(da.coords).issubset(["layer"]):
+            return True
+        if is_scalar(da.values[()]):
+            return True
+        spatial_dims = ["x", "y", "mesh2d_nFaces", "layer"]
+        if not np.any( [coord in spatial_dims for coord in da.coords]):
+            return True
+        return False
+
+    def _mask_spatial_var(self, var: str, mask: GridDataArray)->GridDataArray:
+        da = self.dataset[var]
+        array_mask = mask
```

Comment:
Line 560 to 565 could probably be a separate method. The name of that method then also makes clear what it is doing. Looking at the code it think it is correcting the mask for missing layers
```
def _mask_spatial_var(self, var: str, mask: GridDataArray) -> GridDataArray:
  array_mask = correct_for_missing_layers(da, mask)
  if issubclass(da.dtype.type, numbers.Integral):
      if var == "idomain":
          return da.where(array_mask > 0, other=array_mask)
      else:
           return da.where(array_mask > 0, other=0)
      elif issubclass(da.dtype.type, numbers.Real):
          return da.where(array_mask > 0)
      else:
          raise TypeError(
              f"Expected dtype float or integer. Received instead: {da.dtype}"
            ) 

def correct_for_missing_layers(da, mask):
  array_mask  = mask
   if len(da.dims) < len(da.coords):
            if "layer" in da.coords and "layer" not in da.dims:
                array_mask = mask.sel(layer=da.coords["layer"])
            if "layer" not in da.coords and "layer" in array_mask.coords:
                array_mask = mask.isel(layer=0)

  return array_mask  
```

Reply:
done